rand_xorshift/src/lib.rs

```diff
@@ -28,6 +28,17 @@ use rand_core::{RngCore, SeedableRng, Error, impls, le};
 /// but is very fast. If you do not know for sure that it fits your
 /// requirements, use a more secure one such as `StdRng` or `OsRng`.
 ///
+/// Please be aware how this implementation behaves in pathological cases:
+///
+/// 1. when seeded with zero (i.e. `XorShiftRng::from_seed(0)` is called), it actually uses
+///    `0xBAD_5EED_0BAD_5EED_0BAD_5EED_0BAD_5EED` for the seed. This arbitrary seed is used because
+///    the underlying algorithm can't escape from an all-zero state, and the function is infallible
+///    so it can't signal this by returning an error;
+///
+/// 2. when seeded from an RNG that always returns zeroes, `XorShiftRng::from_rng` will loop
+///    forever. This is again caused by zero being a sticky state for xorshift, and even though the
+///    function is fallible, actually returning an error will break rejection sampling elsewhere.
```

Comment:
@vks, in https://github.com/rust-random/rngs/issues/24#issuecomment-828546664 you said:

>  disagree that we should add special code for pathological RNG implementations that always return 0. Such behavior would break a lot of code in Rand, because rejection sampling is used in many algorithms.

That comment got upvoted by two other contributors, so you probably know what you're talking about, but I don't quite get it. I tried my best to mimic your words here, but if I read this on docs.rs, I'd still be wondering what exactly this behaviour would break. Do you have a suggestion for better phrasing, or can you explain a bit more of what would get broken? I'm happy just being told what part of the rand crate to look at for an example.

Reply:
1. looks good!

I think you can just remove 2., because we don't document for every individual algorithm that it may result in an infinite loop for broken RNG implementations. I would rather improve the `RngCore` docs to state that it is expected that the RNG produces bits uniformly, or it may break many sampling algorithms.

The background is the following: Imagine you want to uniformly generate an integer from the range `0..=6` (for instance, by calling `Rng::gen_range`). Internally, Rand implements this by calling `RngCore::next_u32` and masking three bits. This way, you can still get `0b111`, which is 7 and thus too large. To avoid this, Rand uses rejection sampling, and calls `RngCore::next_u32` until a number different from 7 is obtained (7 is "rejected", because it is outside the intended range). Using this method, all integers in the range `0..=6` have the same probability to be generated. However, if for some reason `RngCore::next_u32` always returns `!0`, this results in an infinite loop.

The same problem applies to many other distributions, which internally use rejection sampling as well.

Reply:
Ah, so I misunderstood the original comment! I thought that *returning an error* will break sampling, but actually it's the broken RNGs that will break stuff. Makes total sense, thanks for the explanation! I'll update this PR to remove the second point, and will take a look at `RngCore`'s docs to see where to fit a warning there.

Reply:
I updated this PR, and submitted https://github.com/rust-random/rand/pull/1118 to add a note to `RngCore`.